server/src/main/java/org/opensearch/index/shard/RemoteStoreRefreshListener.java

```diff
@@ -347,22 +339,70 @@ private boolean isRefreshAfterCommit() throws IOException {
     }
 
-    String uploadSegmentInfosSnapshot(String latestSegmentsNFilename, SegmentInfos segmentInfosSnapshot) throws IOException {
+    /**
+     * Generates the filename for segment info snapshot *
+     * @param commitGeneration commit generation for which filname needs to be generated
+     * @return segment info snapshot filename
+     */
+    private String generateSegmentInfoSnapshotFilename(long commitGeneration) {
+        return SEGMENT_INFO_SNAPSHOT_FILENAME_PREFIX + "__" + commitGeneration;
+    }
+
+    /**
+     * Handles segment info snapshot upload
+     * @param segmentInfoSnapshotFilename remote store filename for segment info snapshot
+     * @param segmentInfosSnapshot segment info used to generate the snapshot
+     * @throws IOException
+     */
+    void uploadSegmentInfosSnapshot(String segmentInfoSnapshotFilename, SegmentInfos segmentInfosSnapshot) throws IOException {
         final long maxSeqNoFromSegmentInfos = indexShard.getEngine().getMaxSeqNoFromSegmentInfos(segmentInfosSnapshot);
 
         Map<String, String> userData = segmentInfosSnapshot.getUserData();
         userData.put(LOCAL_CHECKPOINT_KEY, String.valueOf(maxSeqNoFromSegmentInfos));
         userData.put(SequenceNumbers.MAX_SEQ_NO, Long.toString(maxSeqNoFromSegmentInfos));
         segmentInfosSnapshot.setUserData(userData, false);
+        try {
+            try (IndexOutput indexOutput = storeDirectory.createOutput(segmentInfoSnapshotFilename, IOContext.DEFAULT)) {
+                segmentInfosSnapshot.write(indexOutput);
+            }
+            storeDirectory.sync(Collections.singleton(segmentInfoSnapshotFilename));
+            remoteDirectory.copyFrom(storeDirectory, segmentInfoSnapshotFilename, segmentInfoSnapshotFilename, IOContext.DEFAULT, true);
+            storeDirectory.deleteFile(segmentInfoSnapshotFilename);
+        } catch (Exception e) {
+            tryAndDeleteLocalFile(segmentInfoSnapshotFilename);
+            throw e;
+        }
+    }
 
-        long commitGeneration = SegmentInfos.generationFromSegmentsFileName(latestSegmentsNFilename);
-        String segmentInfoSnapshotFilename = SEGMENT_INFO_SNAPSHOT_FILENAME_PREFIX + "__" + commitGeneration;
-        try (IndexOutput indexOutput = storeDirectory.createOutput(segmentInfoSnapshotFilename, IOContext.DEFAULT)) {
-            segmentInfosSnapshot.write(indexOutput);
+    /**
+     * Handles segment metadata upload.
+     * @param localSegmentsPostRefresh Collection of local segment files after last successful refresh
+     * @param commitGeneration commit generation
+     * @throws IOException
+     */
+    void uploadMetadata(Collection<String> localSegmentsPostRefresh, long commitGeneration) throws IOException {
+        String metadataFilename = remoteDirectory.getMetadataFileName(indexShard.getOperationPrimaryTerm(), commitGeneration);
+        try {
+            remoteDirectory.uploadMetadata(localSegmentsPostRefresh, storeDirectory, metadataFilename);
```

Comment:
IMO, The change of signature in `uploadMetadata` exposes internal implementation of `RemoteSegmentStoreDirectory`. Why don't we handle the deletion in the directory itself?